File: backend/app/models/attendance_report.py
from sqlalchemy import Column, Integer, Float, DateTime, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.database import Base
import uuid
from datetime import datetime


class AttendanceReport(Base):
    __tablename__ = "attendance_reports"

    report_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    class_id = Column(UUID(as_uuid=True), ForeignKey("classes.class_id", ondelete="CASCADE"), nullable=False)
    student_id = Column(UUID(as_uuid=True), ForeignKey("users.user_id", ondelete="CASCADE"), nullable=False)

    total_sessions = Column(Integer, nullable=False, default=0)
    attended_sessions = Column(Integer, nullable=False, default=0)
    late_sessions = Column(Integer, nullable=False, default=0)
    absent_sessions = Column(Integer, nullable=False, default=0)
    left_early_sessions = Column(Integer, nullable=False, default=0)
    reverified_sessions = Column(Integer, nullable=False, default=0)
    attendance_rate = Column(Float, nullable=False, default=0.0)

    generated_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    student = relationship("User", backref="attendance_reports")
    classroom = relationship("Class", backref="attendance_reports")

File: frontend/lib/models/attendance_report.dart
class AttendanceReport {
  final String reportId;
  final String classId;
  final String studentId;
  final int totalSessions;
  final int attendedSessions;
  final int lateSessions;
  final int absentSessions;
  final int leftEarlySessions;
  final int reverifiedSessions;
  final double attendanceRate;
  final String generatedAt;

  AttendanceReport({
    required this.reportId,
    required this.classId,
    required this.studentId,
    required this.totalSessions,
    required this.attendedSessions,
    required this.lateSessions,
    required this.absentSessions,
    required this.leftEarlySessions,
    required this.reverifiedSessions,
    required this.attendanceRate,
    required this.generatedAt,
  });

  factory AttendanceReport.fromJson(Map<String, dynamic> json) {
    return AttendanceReport(
      reportId: json['report_id'],
      classId: json['class_id'],
      studentId: json['student_id'],
      totalSessions: json['total_sessions'] ?? 0,
      attendedSessions: json['attended_sessions'] ?? 0,
      lateSessions: json['late_sessions'] ?? 0,
      absentSessions: json['absent_sessions'] ?? 0,
      leftEarlySessions: json['left_early_sessions'] ?? 0,
      reverifiedSessions: json['reverified_sessions'] ?? 0,
      attendanceRate: (json['attendance_rate'] ?? 0).toDouble(),
      generatedAt: json['generated_at'] ?? '',
    );
  }
}
